### PR TITLE
Sentry.io support to monitor errors

### PR DESCRIPTION
**Pull request by 21h**
At startup check SENTRY_DSN env variable to activate sentry errors logging. Useful for investigation of bugs.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -51,6 +54,17 @@ func init() {
 }
 
 func main() {
+	// Sentry.io logging if user wants to investigate internal errors
+	sentryDSN := os.Getenv("SENTRY_DSN")
+	if sentryDSN != "" {
+		err := sentry.Init(sentry.ClientOptions{
```
Comment (lafriks) [inline main.go]:
```suggestion
		if err := sentry.Init(sentry.ClientOptions{
			Dsn: sentryDSN,
		}); err != nil {
			log.Fatal("Sentry failed to connect: %v", err)
		}
```

**Comment by zeripath** [comment]
Sentry requires that you actually tell it about errors - it's not magic. 

So you would have to next either add multiple specific sentry reporting blocks everywhere that an error is found or, better, just create a sentry logger provider to get it to report an errors.

To then make it really useful requires adding context - this is where the real problems start - you're going to have to place sentry calls almost everywhere to give it context. This might be possible with a macaron.Handler at an appropriate level I guess(?) but it might require changing the macaron router code to actually properly handle breadcrumbs, and the login process etc.

But all of this, placing code at key points in our system, is to provide logging to a proprietary system.

Now... I agree we can do logging better. I agree with Sentry's idea about context - and have been thinking about it quietly for some time (albeit from a python twisted view). However I do not believe that tightly integrating a proprietary system like this in to an open source project like Gitea is necessarily appropriate. 

Now if we can do this in such a way to improve logging in general - then great let's go for it.

**Comment by lafriks** [comment]
@zeripath I have not used it but imho [sentry.io](https://github.com/getsentry/sentry) is open source

**Comment by zeripath** [comment]
@lafriks - ah sorry I missed that you could host a DSN yourself. A quick scan through the documentation implied that you had to pass stuff to their DSN.

File: main.go [hunk 2]
```diff
@@ -51,6 +54,16 @@ func init() {
 }
 
 func main() {
+	// Sentry.io logging if user wants to investigate internal errors
+	sentryDSN := os.Getenv("SENTRY_DSN")
+	if sentryDSN != "" {
+		if err := sentry.Init(sentry.ClientOptions{
```
Comment (lunny) [inline main.go:60]: We still need `repanic: true` ?

**Comment by 21h** [comment]
sentry is a nice choice to get events, it really helps to find problems with code or server itself.

**Comment by alexanderadam** [comment]
> To then make it really useful requires adding context

I just want to add that, besides error tracking, Sentry is also able to [track releases](https://sentry.io/features/releases/)* (I would also love to have [a gitea webhook for that](https://github.com/go-gitea/gitea/issues/7548)) and [deploys](https://docs.sentry.io/api/releases/post-release-deploys/).
So I guess it would be nice as well to see when a certain gitea version was deployed if there is suddenly unwanted behaviour.

And yes, as you correctly pointed out, Sentry is open source and you can simply host it on your own with [the official Docker images for example](https://hub.docker.com/_/sentry/). Similar to gitea.


*I'm not quite sure whether [sentry-go](https://github.com/getsentry/sentry-go) is able to track releases?

**Pull request closed without merging**